functional/stateful-types.md

```diff
@@ -0,0 +1,269 @@
+
+What would really be neat is if it were possible to create a compile-time error
+if it were misused. After all, every user's program contains the invalid call
+order in the logic itself.
+
+In Rust, this is actually possible! The solution is to *change the type* in
+order to enforce the invariants.
+
+How? With a private generic parameter.
+
+Here is what that looks like:
+
+```rust,ignore
+// this is a module to prevent users outside this crate from doing their own impls
+mod state_trait {
+    pub(crate) trait State {}
+
+    pub(crate) struct Init {
+        params: Vec<Param>,
+    }
+    impl State for Init {}
+
+    pub(crate) struct Loaded {
+        scripts: Vec<CompiledScript>,
+    }
+    impl State for Loaded {}
+
+    pub(crate) struct Ready(Vec<CompiledScript>);
+    impl State for Ready {}
+}
+use state_trait::*;
+
+
+struct Interpreter<S: State> {
+    // same fields for execution, but now add:
+    state_data: S,
+}
+
+impl Default for Interpreter<Init> {
+    /* impl does the same thing the old default one did */
+}
+
+impl Interpreter<Init> {
+    /// Set a system prameter.
+    ///
+    /// # Panics
+    /// Will panic if called after a script is executed.
+    ///
+    pub fn set_param(&mut self, p: Param) {
+        /* update the state based on the parameter */
+    }
+
+    /// Initialize this intepreter, disallowing more parameter sets.
+    ///
+    pub fn init(self) -> Interpreter<Loaded> {
+        /* copy all of the parameters into self's members... */
+
+        // return our new initialized interpreter
+        Interpreter {
+            state_data: Loaded { scripts: vec![] },
+            // copy other fields...
+        }
+    }
+}
+
+impl Interpreter<Loaded> {
+    /// Load and compile a script. Returns any errors encountered.
+    ///
+    pub fn compile_script(&mut self, script: &str) -> CompileResult {
+        /* actually do the compile... */
+        Ok(())
+    }
+
+    // Indicates we are done compiling scripts.
+    pub fn ready(self) -> Interpreter<Ready> {
+        /* prepare to actually execute scripts... */
+
+        Interpreter {
+            state_data: Ready(),
+            // copy other fields...
+        }
+    }
+}
+
+impl Interpreter<Ready> {
+    /// Execute all scripts added. Returns any errors encountered.
+    ///
+    /// # Panics
+    /// Will panic if no scripts had been compiled.
+    ///
+    pub fn exec(&mut self) -> ExecutionResult {
+        /* actually run the script... */
+        Ok(())
+    }
+
+    /// Returns to a state where more scripts can be loaded.
+    pub fn reset(self) -> Interpreter<Loaded> {
+        /* clean up any state as needed... */
+
+        Interpreter {
+            state_data: Loaded(self.0),
+            // copy other fields...
+        }
+    }
+}
+```
+
+With this approach, if the user were to make a mistake and set a parameter
+after init:
+
+```rust,ignore
+fn main() {
+    let mut interp = Interpreter::<Init>::default().init();
```

Comment:
Having this turbofish is the downside of what I mentioned. Not ergonomics and what if user does `Interpreter::<Loaded>::default()` instead?

Reply:
The default only being in the init state is precisely what I intended, for that exact reason.

I agree it's clunky, but also I think it's reasonable to recomment. If you don't put the turbofish in at all, I think the current stable has a good suggestion to lead you in the right direction.